### PR TITLE
`azurerm_eventhub_consumer_group` add state migration for resource ID

### LINKED ISSUE #14266: r/eventhub_namespace_authorization_rule: missing state migration
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform (and AzureRM Provider) Version
azurerm version 2.86.0
tested with terraform version 1.0 and 0.15

### Affected Resource(s)

* `azurerm_eventhub_namespace_authorization_rule`

### Terraform Configuration Files

```hcl
# Copy-paste your Terraform configurations here - for large Terraform configs,
# please use a service like Dropbox and share a link to the ZIP file. For
# security, you can also encrypt the files using our GPG public key: https://keybase.io/hashicorp
```

### Debug Output

Error 1 (@tombuildsstuff: this is tracked in #14267)
` Error: parsing "/subscriptions/..../resourceGroups/..../Microsoft.EventHub/namespaces/.../AuthorizationRules/DataWriter": parsing segment "microsoftRelay": expected the segment "Microsoft.EventHub" to be "Microsoft.Relay"`

Error 2
`Error: parsing "/subscriptions/.../resourceGroups/.../providers/Microsoft.EventHub/namespaces/.../eventhubs/.../consumergroups/dashboard": parsing segment "staticConsumerGroups": expected the segment "consumergroups" to be "consumerGroups"`

### Steps to Reproduce

Rerun a pipline with existing eventhub or azurerm_eventhub_namespace_authorization_rule after upgdate from 2.85 to 2.86

1. `terraform apply`

### References

* #0000

### PR DESCRIPTION
**Pull request by stephybun**
Fixes #14266

### CODE REVIEW

File: internal/sdk/resource.go [hunk 1]
```diff
@@ -63,7 +63,16 @@ type Resource interface {
 	IDValidationFunc() pluginsdk.SchemaValidateFunc
 }
 
-// TODO: ResourceWithStateMigration
+type ResourceWithStateMigration interface {
+	Resource
+	StateUpgradeDataFunc() StateUpgradeData
```
Comment (tombuildsstuff) [inline internal/sdk/resource.go:68]: could we make this:

```suggestion
	StateUpgraders() StateUpgradeData
```

File: internal/services/eventhub/eventhub_consumer_group_resource.go [hunk 2]
```diff
@@ -208,3 +210,10 @@ func (r ConsumerGroupResource) ModelObject() interface{} {
 func (r ConsumerGroupResource) IDValidationFunc() pluginsdk.SchemaValidateFunc {
 	return consumergroups.ValidateConsumerGroupID
 }
+
+func (r ConsumerGroupResource) StateUpgradeDataFunc() sdk.StateUpgradeData {
+	return sdk.StateUpgradeData{
+		SchemaVersion: 1,
+		Upgraders:     map[int]pluginsdk.StateUpgrade{0: migration.ConsumerGroupsV0ToV1{}},
```
Comment (tombuildsstuff) [inline internal/services/eventhub/eventhub_consumer_group_resource.go:217]: can we make this:

```suggestion
		Upgraders:     map[int]pluginsdk.StateUpgrade{
			0: migration.ConsumerGroupsV0ToV1{},
		},
```

(not a biggie, it's just easier to read)

File: internal/services/eventhub/migration/consumer_group.go [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+				consumerGroupSegment = value
+				break
+			}
+		}
+
+		if consumerGroupSegment == "" {
+			return rawState, fmt.Errorf("couldn't find the `consumerGroups` segment in the old resource id %q", oldId)
+		}
+
+		newId := consumergroups.NewConsumerGroupID(oldId.SubscriptionID, oldId.ResourceGroup, oldId.Path["namespaces"], oldId.Path["eventhubs"], consumerGroupSegment)
```
Comment (tombuildsstuff) [inline internal/services/eventhub/migration/consumer_group.go:41]: the parser has an Insensitively function, so we can make this:

```suggestion
		oldId := rawState["id"].(string)
		parsed, err := ParseConsumerGroupIDInsensitively(oldId)
		if err != nil {
			return fmt.Errorf("parsing existing Consumer Group ID %q: %+v", oldId, err)
		}
		
		newId := parsed.ID()
```

**Review by tombuildsstuff — CHANGES_REQUESTED** [review]
hey @stephybun 

Thanks for this PR - I've taken a look through and left a few minor comments inline but if we can fix those up then this should otherwise be good to go 👍 

Thanks!

File: internal/services/eventhub/migration/consumer_group.go (hunk 3 above)
Comment (stephybun) [reply internal/services/eventhub/migration/consumer_group.go]: ohh nifty!

**Comment by stephybun** [comment]
Thanks @tombuildsstuff, I applied your suggestions.

**Review by tombuildsstuff — APPROVED** [review]
lgtm 👍

**Pull request merged**